### PR TITLE
feat(iot): support Thing L2 construct

### PR DESCRIPTION
**Pull request by yamatatsu**
This PR add Thing L2 construct. Issue #18872 

If we close this issue, it is also needed to support ThingGroup and ThingType.
So I didn't use _fix_ keyword for above issue link.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iot/README.md [hunk 1]
```diff
@@ -83,3 +83,16 @@ new iot.TopicRule(this, 'TopicRule', {
 ```
 
 See also [@aws-cdk/aws-iot-actions](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-iot-actions-readme.html) for other actions.
+
+## Thing
+
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-iot/README.md:88]: can we add a sentence or two on why/when to use a Thing?

File: packages/@aws-cdk/aws-iot/README.md [hunk 2]
```diff
@@ -83,3 +83,16 @@ new iot.TopicRule(this, 'TopicRule', {
...
+```ts
+new iot.Thing(this, 'Thing', {
+  thingName: 'my-thing', // optional property, default is a generated name
+  attributes: { // optional property, default is none
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-iot/README.md:92]: why/when to use attributes?

**Review by kaizencc — COMMENTED** [review]
Hi @yamatatsu! Sorry for the delay in getting a review on this! In general this looks fairly straightforward and I'm happy to approve. But, given that the issue is asking for `ThingGroup` and `ThingType`, I have a few questions (pardon my lack of knowledge in iot)

- on its own, is a `Thing` construct useful? the readme doesn't really go in depth about why/when to use it
- what is your proposed game plan for supporting `ThingGroup` and `ThingType` given the lack of cfn support? custom resources?

Lets lay out the long term plan, and then if you're willing, I'm happy to help review any additional PRs as well.

File: packages/@aws-cdk/aws-iot/README.md (hunk 2 above)
Comment (yamatatsu) [reply packages/@aws-cdk/aws-iot/README.md:92]: Usually `attributes` store about static information about the thing, such as product name, serial number, lot number or manufacturer. And it is used for search things on web console and cli.

File: packages/@aws-cdk/aws-iot/README.md (hunk 1 above)
Comment (yamatatsu) [reply packages/@aws-cdk/aws-iot/README.md:88]: Yes, we can. I'll write it in this PR!

**Comment by yamatatsu** [comment]
@kaizencc 
Thank you for your review!

> on its own, is a Thing construct useful? the readme doesn't really go in depth about why/when to use it

Right. I'll explain it in README!

> what is your proposed game plan for supporting ThingGroup and ThingType given the lack of cfn support? custom resources?

Yes, I believe I will be able to create the ThingGroup L2 construct with custom resources.

But ThingType is difficult because for deleting a ThingType, there are two rules:

1. It is needed to mark a ThingType as depricated to delete the ThingType.
1. It is needed to wait 5 minutes after it is marked as deprecated.

So I am going to create a ThingGroup first and think about ThingType after that.

**Comment by yamatatsu** [comment]
If it is better, I'll create the RFC issue. is it?

**Comment by kaizencc** [comment]
Hi @yamatatsu! If you would _prefer_ to get additional help on design, then we can definitely go the RFC route. However, I also recognize that you have contributed often to the iot modules and are also a `star-contributor` :). There's an additional level of trust gained from that, so if you feel confident in your design then I'm willing to work with you in PR form. This particular PR can go in once the small nits are addressed.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Putting in changes requested to reflect status.

**Pull request closed without merging**